xapian-core/cluster/cluster.cc

```diff
@@ -149,24 +150,109 @@ ClusterSet::cluster_size(clusterid cid) {
 	LOGCALL(API, doccount, "ClusterSet::cluster_size(clusterid cid)", cid);
 	map<clusterid, vector<Document> >::iterator i;
 	i = clusters.find(cid);
-	//Add error handling here
-	if(i != clusters.end())
-		return ((i->second).size());	
+	if(i == clusters.end())
+		throw RangeError("The mentioned clusterid was out of range", 103);
+	return ((i->second).size());	
+}
+
+DocumentSet
+ClusterSet::get_cluster(clusterid cid) {
+	LOGCALL(API, vector<Document>, "ClusterSet::get_cluster(clusterid cid)", cid);
+	map<clusterid, vector<Document> >::iterator i;
+	i = clusters.find(cid);
+	if(i == clusters.end())
+		throw RangeError("The mentioned clusterid was out of range", 103);
+	vector<Document> docs;
+	docs = i->second;
+	DocumentSet docset;
+	for(vector<Document>::iterator k = docs.begin(); k != docs.end(); k++)
+		docset.add_document(*k);
+	return docset;
+}
+
+void
+ClusterSet::add_document(clusterid cid, Document &doc) {
+	LOGCALL_VOID(API, "ClusterSet::add_document(clusterid cid, Document doc)", cid | doc);
+	clusters[cid].push_back(doc);
+}
+
+DocumentSet
+ClusterSetIterator::get_cluster() {
+	LOGCALL(API, DocumentSet, "ClusterSetIterator::get_cluster()", "");
+	return cset.get_cluster(index);
 }
 
 ClusterSetIterator
 ClusterSet::begin() {
-	LOGCALL(API, ClusterSetIterator, "ClusterSet::begin()","");
+	LOGCALL(API, ClusterSetIterator, "ClusterSet::begin()", "");
 	return ClusterSetIterator(*this, 0);
 }
 
 ClusterSetIterator
 ClusterSet::end() {
-	LOGCALL(API, ClusterSetIterator, "ClusterSet::end()","");
+	LOGCALL(API, ClusterSetIterator, "ClusterSet::end()", "");
 	return ClusterSetIterator(*this, num_of_clusters());
 }
 
 ClusterSetIterator
 ClusterSet::operator[](doccount i) {
 	return ClusterSetIterator(*this, i);
 }
+
+void
+DocumentSet::add_document(Document &doc) {
+	LOGCALL_VOID(API, "DocumentSet::add_document()", doc);
+	docs.push_back(doc);
+}
+
+Document
+DocumentSet::operator[](doccount i) {
+	Document d;
+	d = docs[i];
+	return d;
+}
+
+DocumentSetIterator
+DocumentSet::begin() {
+	LOGCALL(API, DocumentSetIterator, "DocumentSet::begin()", "");
+	return DocumentSetIterator(*this, 0);
+}
+
+DocumentSetIterator
+DocumentSet::end() {
+	LOGCALL(API, DocumentSetIterator, "DocumentSet::end()", "");
+	return DocumentSetIterator(*this, size());
+}
+
+Document
+DocumentSetIterator::get_document() {
+	LOGCALL(API, Document, "DocumentSetIterator::get_document()", "");
+	Document d;
+	d = docs[index];
+	return d;
+}
+
+RoundRobin::~RoundRobin()
```

Comment:
I'd probably put the `RoundRobin` implementation in its own file, to keep it separate from the "mechanics" of the clustering system that are separate to any algorithm. (Otherwise, the next algorithm will also be in this file, and it'll become harder and harder to find the bit you need.)